test/e2e/plain-v0_provisioner_test.go

```diff
@@ -192,3 +194,188 @@ var _ = Describe("plain-v0 provisioner", func() {
 		})
 	})
 })
+
+var _ = Describe("plain-v0 provisioner bundleinstance", func() {
+	When("a BundleInstance targets a valid Bundle", func() {
+		var (
+			bundle *v1alpha1.Bundle
+			bi     *v1alpha1.BundleInstance
+		)
+		BeforeEach(func() {
+			By("creating the testing Bundle resource")
+			bundle = &v1alpha1.Bundle{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "olm-crds",
+				},
+				Spec: v1alpha1.BundleSpec{
+					ProvisionerClassName: plainProvisionerID,
+					Image:                "quay.io/tflannag/olm-plain-bundle:olm-crds-v0.20.0",
+				},
+			}
+			err := c.Create(ctx, bundle)
+			Expect(err).To(BeNil())
+
+			bi = &v1alpha1.BundleInstance{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "olm-crds",
+				},
+				Spec: v1alpha1.BundleInstanceSpec{
+					ProvisionerClassName: plainProvisionerID,
+					BundleName:           bundle.GetName(),
+				},
+			}
+			err = c.Create(ctx, bi)
+			Expect(err).To(BeNil())
+		})
+		AfterEach(func() {
+			By("deleting the testing Bundle resource")
+			Eventually(func() error {
+				return client.IgnoreNotFound(c.Delete(ctx, bundle))
+			}).Should(Succeed())
+
+			By("deleting the testing BundleInstance resource")
+			Eventually(func() error {
+				return c.Delete(ctx, bi)
+			}).Should(Succeed())
+		})
+
+		It("should rollout the bundle contents successfully", func() {
+			By("eventually writing a successful installation state back to the bundleinstance status")
+			Eventually(func() bool {
+				if err := c.Get(ctx, client.ObjectKeyFromObject(bi), bi); err != nil {
+					return false
+				}
+				if bi.Status.InstalledBundleName != bundle.GetName() {
+					return false
+				}
+				existing := meta.FindStatusCondition(bi.Status.Conditions, "Installed")
+				if existing == nil {
+					return false
+				}
+				expected := metav1.Condition{
+					Type:    "Installed",
+					Status:  metav1.ConditionStatus(corev1.ConditionTrue),
+					Reason:  "InstallationSucceeded",
+					Message: "",
+				}
+				return conditionsSemanticallyEqual(expected, *existing)
+			}).Should(BeTrue())
+
+			By("eventually reseting a bundle lookup failure when the targeted bundle has been deleted")
+			Eventually(func() error {
+				return c.Delete(ctx, bundle)
+			}).Should(Succeed())
+
+			Eventually(func() bool {
+				if err := c.Get(ctx, client.ObjectKeyFromObject(bi), bi); err != nil {
+					return false
+				}
+				if bi.Status.InstalledBundleName != "" {
+					return false
+				}
+				existing := meta.FindStatusCondition(bi.Status.Conditions, "Installed")
+				if existing == nil {
+					return false
+				}
+				expected := metav1.Condition{
```

Comment:
In this test case, when the bundle was installed, BI was created, and then the bundle was removed, we show this status condition. But the BI is on cluster right, so the objects in the bundle are still on-cluster as well. The underlying storage of those objects was removed. This condition may be confusing because "`Installed: False`" suggests that the application is not installed, but it still is?

Reply:
That's a good call out - I was fairly confident that you were right, and that we still left the resources on-cluster as the owner references for those resources _should_ be the BundleInstance and not the unpacked Bundle object, but I double checked this assumption locally anyways:

```bash
$ k get bundleinstances olm-crds -o yaml
apiVersion: core.rukpak.io/v1alpha1
kind: BundleInstance
metadata:
  name: olm-crds
  resourceVersion: "115145"
  ...
spec:
  bundleName: olm-v0.20.0-crds
  provisionerClassName: core.rukpak.io/plain-v0
status:
  conditions:
  - lastTransitionTime: "2022-03-04T17:17:20Z"
    message: Bundle.core.rukpak.io "olm-v0.20.0-crds" not found
    reason: BundleLookupFailed
    status: "False"
    type: Installed
$ k get crds
NAME                                          CREATED AT             LABELS
bundleinstances.core.rukpak.io                2022-03-03T22:38:42Z   <none>
bundles.core.rukpak.io                        2022-03-03T22:38:42Z   <none>
catalogsources.operators.coreos.com           2022-03-04T17:16:52Z   app.kubernetes.io/managed-by=Helm,core.rukpak.io/owner-kind=BundleInstance,core.rukpak.io/owner-name=olm-crds
clusterserviceversions.operators.coreos.com   2022-03-04T17:16:52Z   app.kubernetes.io/managed-by=Helm,core.rukpak.io/owner-kind=BundleInstance,core.rukpak.io/owner-name=olm-crds
installplans.operators.coreos.com             2022-03-04T17:16:52Z   app.kubernetes.io/managed-by=Helm,core.rukpak.io/owner-kind=BundleInstance,core.rukpak.io/owner-name=olm-crds
olmconfigs.operators.coreos.com               2022-03-04T17:16:52Z   app.kubernetes.io/managed-by=Helm,core.rukpak.io/owner-kind=BundleInstance,core.rukpak.io/owner-name=olm-crds
operatorconditions.operators.coreos.com       2022-03-04T17:16:52Z   app.kubernetes.io/managed-by=Helm,core.rukpak.io/owner-kind=BundleInstance,core.rukpak.io/owner-name=olm-crds
operatorgroups.operators.coreos.com           2022-03-04T17:16:52Z   app.kubernetes.io/managed-by=Helm,core.rukpak.io/owner-kind=BundleInstance,core.rukpak.io/owner-name=olm-crds
operators.operators.coreos.com                2022-03-04T17:16:52Z   app.kubernetes.io/managed-by=Helm,core.rukpak.io/owner-kind=BundleInstance,core.rukpak.io/owner-name=olm-crds
subscriptions.operators.coreos.com            2022-03-04T17:16:52Z   app.kubernetes.io/managed-by=Helm,core.rukpak.io/owner-kind=BundleInstance,core.rukpak.io/owner-name=olm-crds
```

Maybe this is something we can iron out in #99, or tackle during the refactor of the BundleInstance controller? Should I just add a TODO comment for tracking this behavior?

Reply:
I think #99 is slightly different, it's more of a feature to be honest. I think a TODO comment is fine and we can look at it during the refactor. Essentially, we have to figure out how to describe the state of the BI in the test.

Does the BI attempt to redeploy resources that pertain to a particular installation if an object is removed? Looking at the code, it looks like the BI controller sets up dynamic watches for all objects associated with the install, so that would result in a requeue of the BI in case of a deletion event, but I'm not sure if we actually recreate objects. If so, then deleting the bundle would potentially put the BI in a bad state if it can't recreate objects as expected. If on the other hand the BI creates the objects from the bundle once and then does not manage them then removing the Bundle isn't as impactful. The first case seems like more kube-like behavior however. If that's the case then I think we would need to have a custom status for this condition like `BundleRemoved` indicating that the installation was successful but the underlying Bundle is no longer on the cluster. As long as we do the most intuitive thing it shouldn't confuse users.

Reply:
We need a concept of the BI's current "release". That release becomes the thing that the BI controller uses its dynamic watches to ensure that the underlying release objects aren't modified or deleted.

A release needs to be pivoted. During a pivot, I'd expect that we'd maintain the state of both the old and new releases. If the new release fails to be applied or fails to achieve a successful state within a certain period of time, we need the old release to remain active.

From a BI status standpoint, what I think this boils down to is that we need to differentiate between the state of what's active and the state of what is currently being attempted.